src/service/cached-xhr-impl.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../log';
+import {fromClass} from '../service';
+import {
+  parseUrl,
+  removeFragment,
+} from '../url';
+import {getPath, hasOwn} from '../utils/object';
+import {isArray, isObject, isFormData} from '../types';
+import {Cache} from '../utils/cache';
+import {Xhr} from './xhr-impl';
+
+/**
+ * The "init" argument of the Fetch API. Currently, only "credentials: include"
+ * is implemented.  Note ampCors with explicit false indicates that
+ * __amp_source_origin should not be appended to the URL to allow for
+ * potential caching or response across pages.
+ *
+ * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+ *
+ * @typedef {{
+ *   body: (!Object|!Array|undefined|string),
+ *   credentials: (string|undefined),
+ *   headers: (!Object|undefined),
+ *   method: (string|undefined),
+ *   requireAmpResponseSourceOrigin: (boolean|undefined),
+ *   ampCors: (boolean|undefined)
+ * }}
+ */
+let FetchInitDef;
+
+
+/** @private @const {!Array<function(*):boolean>} */
+const allowedJsonBodyTypes_ = [isArray, isObject];
+
+
+/**
+ * A service that polyfills Fetch API for use within AMP.
+ *
+ * @package Visible for type.
+ * @visibleForTesting
+ */
+export class CachedXhr extends Xhr {
+
+  /**
+   * @param {!Window} win
+   * @param {number=} cacheSize
+   */
+  constructor(win, cacheSize) {
+    super(win);
+
+    /** @const {!Cache} */
+    this.fragmentCache_ = new Cache(cacheSize);
+  }
+
+  /**
+   * Fetches and constructs JSON object based on the fetch polyfill.
+   *
+   * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+   *
+   * See `fetchAmpCors_` for more detail.
+   *
+   * @param {string} input
+   * @param {?FetchInitDef=} opt_init
+   * @return {!Promise<!JSONType>}
+   * @override
+   */
+  fetchJson(input, opt_init) {
```

Comment:
There's a lot of duplication with `Xhr#fetchJson` here. I think we can simplify down based on:

1. Any call with `opt_init` is un-cacheable
  - Removes crazy `simpleSerialize`
  - Handles `POST` case

Actually, that's it. 😉